doc/man/metacity-message.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\" -----
.\" This file was confirmed to be licenced under the GPL
.\" by its author and copyright holder, Akira TAGOH, on June 1st 2008:
.\"
.\" > I'm comfortable with DFSG-free. that sounds great if you think it's
.\" > useful and worth containing it in upstream.
.\" ...
.\" > Right I know. any licenses that is DFSG-free, I'm ok with whatever,
.\" > since I have contributed that for Debian. so GPL is no problem for me.
.\" -----
.TH METACITY\-MESSAGE 1 "16 May 2013"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
METACITY\-MESSAGE \- a command to send a message to Metacity
.SH SYNOPSIS
.B METACITY\-MESSAGE
[ restart | reload\-theme | enable\-keybindings | disable\-keybindings |
enable\-mouse\-button\-modifiers | disable\-mouse\-button\-modifiers | toggle\-verbose ]
.SH DESCRIPTION
This manual page documents briefly the
.B metacity\-message\fP.
This manual page was written for the Debian distribution
because the original program does not have a manual page.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBmetacity\-message\fP send a specified message to \fBmetacity\fP(1).
.SH OPTIONS
.TP
.B restart
Restart \fBmetacity\fP(1) which is running.
.TP
.B reload-theme
Reload a theme which is specified on gsettings database.
.TP
.B enable-keybindings
Enable all of keybindings which is specified on gsettings database.
.TP
.B disable-keybindings
Disable all of keybindings which is specified on gsettings database.
.TP
.B enable-mouse-button-modifiers
Enable move/resize operations while pressing a modifier key
.TP
.B disable-mouse-button-modifiers
Disable move/resize operations while pressing a modifier key
.TP
.B toggle-verbose
Turn debug messages on or off
.SH SEE ALSO
.BR metacity (1)
.SH AUTHOR
This manual page was written by Akira TAGOH <tagoh@debian.org>,
for the Debian GNU/Linux system (but may be used by others).


NEWS
Version 3.14.1
==============
- Fix random bug when windows opens in fullscreen mode. (Alberts Muktupāvels)
- Fix identification of CSD windows when checking whether to force fullscreen.
  (Owen W. Taylor)
- Fix uninitialized variables (Vincent Untz)
- Updated translations:
  Andika Triwidada (id), Piotr Drąg (pl)

Version 3.14.0
==============
- Theme format version updated to 3.5, mostly patches from mutter. (Owen W. Taylor,
  Florian Müllner, Jasper St. Pierre)
- Optionally attach modal dialogs. (Maxim Ermilov)
  [This was added as dependency for updating theme format version and is not 
  available for use in metacity]
- Updated *.doap file. (Alberts Muktupāvels, Olav Vitters)
- Added support for _GTK_FRAME_EXTENTS. (Again based on mutter)
- Added support for _GTK_THEME_VARIANT. (Patches from mutter, mostly work
  by Florian Müllner)
- Shadows won't be added to maximized windows. (Alberts Muktupāvels)
- Fixed some deprecated warnings. (Alberts Muktupāvels)
- Ignore adding a window if it's already present. (hordepfo)
- Some other fixes/changes adapted from mutter.
- Updated translations:
  Andika Triwidada (id), Daniel Mustieles (es), Enrico Nicoletto (pt_BR),
  Gábor Kelemen (hu), genodeftest (de), Inaki Larranaga Murgoiti (eu),
  Marek Černocký (cs), MarMav (el), Piotr Drąg (pl), Rafael Ferreira (pt_BR).

Version 3.12.0
==============
 - Port GTK+ 2 to GTK+ 3 (Mutter developers)
 - Fix title bar button background (Nicolas Desfontaine)
 - Side-by-side window tiling (Maarten Baert)
 - Window placement patch (Chad Glendenin, http://chad.glendenin.com/metacity/patch.html)
 - Add UXD shadows and borders (Sam Spilsbury)
 - Add error handling to glibtop_get_proc_uid (Tomaž Šolc)
 - Cut with fading the Bar title when it's longer than the decoration bar (Marco Trevisan)
 - Don't show the ws switcher if we only have one (Didier Roche)
 - Update meta_show_dialog for gtk's removal of --screen (Dan Winship)
 - Add drag and drop support when using alt + tab (Matthias Clasen)
 - Fix new window focus (https://bugzilla.gnome.org/show_bug.cgi?id=534752)
 - Compositing manager is enabled by default now
 - Fix screen flickering on session start (https://bugzilla.gnome.org/show_bug.cgi?id=658036)

Translations:
    Piotr Drąg (pl)
    Lasse Liehu (fi)
    Wylmer Wang (zh_CN)
    Rafael Ferreira (pt_BR)
    Benjamin Steinwender (de)
    Мирослав Николић (sr, sr@latin)
    Rūdolfs Mazurs (lv)
    Joe Hansen (da)
    Milo Casagrande (it)
    Balázs Úr (hu)
    Alexandre Franke (fr)
    Aurimas Černius (lt)
    Christian Kirbach (de)
    Andika Triwidada (id)
    Matej Urbančič (sl)
    Daniel Mustieles (es)
    Marek Černocký (cs)
    Gil Forcada (ca)
    Fran Diéguez (gl)
    Gheyret Kenji (ug)

Version 2.34.13
===============

- Fix minimization problem of iconified windows (Florian Müllner, Daniel Drake)

Translations:
    Rūdolfs Mazurs (lv)
    Rajesh Ranjan (hi)
    Nguyễn Thái Ngọc Duy (vi)
    Theppitak Karoonboonyanan (th)
    Piotr Drąg (pl)

Version 2.34.8
==============
 - Port themes documentation to yelp-tools (Florian Müllner)

Translations:
    Mattias Põldaru (et)
    Nilamdyuti Goswami (as)
    Fran Diéguez (gl)

Version 2.34.5
==============
 - Fix memory leak (Pavel Vasin)
 - Add disable-mouse-button-modifiers message (Daniel Drake)

Translations:
    Daniel Korostil (uk)
    Jiro Matsuzawa (ja)
    Reşat SABIQ (crh)
    Aurimas Černius (lt)
    Carles Ferrando (ca@valencia)
    Arash Mousavi (fa)
    Kristjan SCHMIDT (eo)
    Sandeep Sheshrao Shedmake (mr)
    Jordi Serratosa (ca)
    Tom Tryfonidis (el)
    Praveen Illa (te)
    Luca Ferretti (it)
    Andika Triwidada (id)
    Daniel Mustieles (es)
    Nilamdyuti Goswami (as)


Version 2.34.3
==============

 - Use the correct translation domain in keybinding definitions (Ryan Lortie, Florian Müllner)

Translations:
    YunQiang Su (zh_CN)
    Abderrahim Kitouni (ar)
    Marek Černocký (cs)
    David Planella (ca)
    Sasi Bhushan (te)
    A S Alam (pa)
    Duarte Loreto (pt)
    Antonio Fernandes C. Neto (pt_BR)
    Anita Reitere (lv)
    Bruce Cowan (en_GB)
    Daniel Nylander (sv)
    Timo Jyrinki (fi)
    Nilamdyuti Goswami (as)
    Piotr Drąg (pl)
    Claude Paroz (fr)
    Seong-ho Cho (ko)
    Inaki Larranaga Murgoitio (eu)
    Mario Blättermann (de)
    Nguyễn Thái Ngọc Duy (vi)
    Gabor Kelemen (hu)
    Daniel Mustieles (es)
    Dr.T.Vasudevan (ta)
    Kasia Bondarava (be)
    Alexander Shopov (bg)
    Ask H. Larsen (da)
    Мирослав Николић (sr)
    Yuri Myasoedov (ru)
    Mattias Põldaru (et)


Version 2.34.2
==============

 - Port preferences to GSettings (Florian Müllner, Milan Bouchet-Valat, Jason Siefken)
 - Remove deprecated GTK+ symbols and use accessor functions (Florian Müllner)
 - Include the default window icon (Matthias Clasen)
 - Various small build system changes (Javier Jardón, Piotr Drąg, Florian Müllner, Rico Tzschichholz)

Translations:
    Abduxukur Abdurixit (ug)
    Kjartan Maraas (nb)
    Theppitak Karoonboonyanan (th)
    Ihar Hrachyshka (be)
    Aurimas Černius (lt)
    Daniel Mustieles (es)
    Jiro Matsuzawa (ja)
    Fabrício Godoy (pt_BR)
    krishnababu k (te)
    OKANO Takayoshi (ja)
    Mario Blättermann (de)
    Bruce Cowan (en_GB)
    Mattias Põldaru (et)
    Fran Diéguez (gl)
    Marek Černocký (cs)
    Gil Forcada (ca)
    Piotr Drąg (pl)
    Matej Urbančič (sl)
    Bruno Brouard (fr)
    Daniel Nylander (sv)
    Praveen Illa (te)
    Wouter Bolsterlee (nl)
    Antonio Fernandes C. Neto (pt_BR)
    Yaron Shahrabani (he)
    Xandru Armesto (ast)
    Alexander Shopov (bg)
    Chao-Hsiung Liao (zh_HK, zh_TW)

==============
Version 2.34.1
==============

 - Fix crash in meta_display_get_above_tab_keycode (Adam Jackson)
 - Disable keybindings message should ungrab keys (Daniel Drake)
 - Remove --with-gtk configure flag as metacity only builds against gtk2 (Colin
   Walters)
 - Build fix for FreeBSD: Include sys/wait.h as described in POSIX specs (Alexis Ballier)

src/tools/metacity-message.c
/* Metacity send-magic-messages app */

/*
 * Copyright (C) 2002 Havoc Pennington
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>
#include <gtk/gtk.h>
#include <gdk/gdkx.h>
#include <stdlib.h>
#include <string.h>

#include <libintl.h>
#define _(x) dgettext (GETTEXT_PACKAGE, x)
#define N_(x) x


static void
send_restart (void)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = GDK_DISPLAY_XDISPLAY (gdk_display_get_default ());
  xev.xclient.window = gdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()),
                                          "_METACITY_RESTART_MESSAGE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = 0;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()),
              gdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()));
  XSync (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()), False);
}

static void
send_reload_theme (void)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = GDK_DISPLAY_XDISPLAY (gdk_display_get_default ());
  xev.xclient.window = gdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()),
                                          "_METACITY_RELOAD_THEME_MESSAGE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = 0;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()),
              gdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()));
  XSync (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()), False);
}

static void
send_set_keybindings (gboolean enabled)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = GDK_DISPLAY_XDISPLAY (gdk_display_get_default ());
  xev.xclient.window = gdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()),
                                          "_METACITY_SET_KEYBINDINGS_MESSAGE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = enabled;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()),
              gdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()));
  XSync (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()), False);
}

static void
send_set_mousemods (gboolean enabled)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = GDK_DISPLAY_XDISPLAY (gdk_display_get_default ());
  xev.xclient.window = gdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()),
                                          "_METACITY_SET_MOUSEMODS_MESSAGE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = enabled;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()),
              gdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()));
  XSync (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()), False);
}

#ifdef WITH_VERBOSE_MODE
static void
send_toggle_verbose (void)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = GDK_DISPLAY_XDISPLAY (gdk_display_get_default ());
  xev.xclient.window = gdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()),
                                          "_METACITY_TOGGLE_VERBOSE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = 0;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()),
              gdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()));
  XSync (GDK_DISPLAY_XDISPLAY (gdk_display_get_default ()), False);
}
#endif

static void
usage (void)
{
  g_printerr (_("Usage: %s\n"),
              "metacity-message (restart|reload-theme|enable-keybindings|disable-keybindings|enable-mouse-button-modifiers|disable-mouse-button-modifiers|toggle-verbose)");
  exit (1);
}

int
main (int argc, char **argv)
{
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");

  gtk_init (&argc, &argv);

  if (argc < 2)
    usage ();

  if (strcmp (argv[1], "restart") == 0)
    send_restart ();
  else if (strcmp (argv[1], "reload-theme") == 0)
    send_reload_theme ();
  else if (strcmp (argv[1], "enable-keybindings") == 0)
    send_set_keybindings (TRUE);
  else if (strcmp (argv[1], "disable-keybindings") == 0)
    send_set_keybindings (FALSE);
  else if (strcmp (argv[1], "enable-mouse-button-modifiers") == 0)
    send_set_mousemods (TRUE);
  else if (strcmp (argv[1], "disable-mouse-button-modifiers") == 0)
    send_set_mousemods (FALSE);
  else if (strcmp (argv[1], "toggle-verbose") == 0)
    {
#ifndef WITH_VERBOSE_MODE
      g_printerr (_("Metacity was compiled without support for verbose mode\n"));
      return 1;
#else
      send_toggle_verbose ();
#endif
    }
  else
    usage ();

  return 0;
}

